coloc/coloc_runner.py

```diff
@@ -0,0 +1,251 @@
+
+1) Identify eGenes using FDR < fdr_threshold (default: 0.05)
+2) Extract the SNP GWAS data for the cis-window (gene +/- window; default: 100kB)
+3) Run coloc for each eGene (if the SNP GWAS data has at least one variant with pval <5e-8)
+4) Write the results to a TSV file
+
+analysis-runner --dataset "bioheart" \
+    --description "Run coloc for eGenes identified by SAIGE-QTL analysis" \
+    --access-level "test" \
+    --memory='8G' \
+    --image "australia-southeast1-docker.pkg.dev/analysis-runner/images/driver:d4922e3062565ff160ac2ed62dcdf2fba576b75a-hail-8f6797b033d2e102575c40166cf0c977e91f834e" \
+    --output-dir "saige-qtl/bioheart_n990_and_tob_n1055/241004_n100/" \
+    coloc/coloc_runner.py \
+    --snp-gwas-file=gs://cpg-bioheart-test/str/gwas_catalog/gcst/gcst-gwas-catalogs/ibd_EAS_EUR_SiKJEF_meta_IBD.tsv \
+    --pheno-output-name="ibd_liu2023" \
+    --celltypes "B_naive"
+
+"""
+
+import click
+import pandas as pd
+
+import hailtop.batch as hb
+
+from cpg_utils import to_path
+from cpg_utils.hail_batch import get_batch, image_path, output_path
+
+
+def coloc_runner(gwas, eqtl_file_path, celltype, coloc_results_file):
+    import rpy2.robjects as ro
+    from rpy2.robjects import pandas2ri
+
+    ro.r('library(coloc)')
+    ro.r('library(tidyverse)')
+
+    with (ro.default_converter + pandas2ri.converter).context():
+        gwas_r = ro.conversion.get_conversion().py2rpy(gwas)
+    ro.globalenv['gwas_r'] = gwas_r
+    ro.r(
+        '''
+    gwas_r = gwas_r %>% select(beta, varbeta, position,snp)
+    gwas_r = gwas_r %>% filter((beta!=0) | (varbeta!=0))
+    gwas_r = gwas_r %>% distinct(snp, .keep_all = TRUE)
+    gwas_r = gwas_r%>% as.list()
+    gwas_r$type = 'cc'
+
+    ''',
+    )
+    eqtl = pd.read_csv(
+        eqtl_file_path,
+        sep='\t',
+    )
+    eqtl['chr'] = eqtl['CHR'].apply(lambda chr: f'chr{chr}')
+    eqtl['beta'] = eqtl['BETA']
+    eqtl['se'] = eqtl['SE']
+    eqtl['position'] = eqtl['POS']
+    eqtl['snp'] = eqtl['MarkerID'].apply(lambda snp: 'chr' + snp.replace(':', '_'))
+    # while I figure out if it's easy to extract sdY, give N and MAF instead
+    # https://chr1swallace.github.io/coloc/articles/a02_data.html#what-if-i-dont-have-sdy
+    eqtl['MAF'] = eqtl['AF_Allele2'].apply(lambda af: min(af, (1 - af)))
+    gene = eqtl_file_path.split('/')[-1].replace(f'{celltype}_', '').replace('_cis', '')
+    with (ro.default_converter + pandas2ri.converter).context():
+        eqtl_r = ro.conversion.get_conversion().py2rpy(eqtl)
+    ro.globalenv['eqtl_r'] = eqtl_r
+    ro.globalenv['gene'] = gene
+    ro.r(
+        '''
+    eqtl_r = eqtl_r %>% filter(!is.na(beta))
+    eqtl_r = eqtl_r %>% distinct(snp, .keep_all = TRUE)
+    eqtl_r$varbeta = eqtl_r$se**2
+    eqtl_r$position = eqtl_r$pos
+    eqtl_r = eqtl_r %>% select(beta, varbeta, position, snp, N, MAF)
```

Comment:
Just confirming, N is a field that's provided for each site in your eQTL input file?

Reply:
Yep, N is one out of outputs of SAIGE-QTL! For the test example file I am using:
<img width="1310" alt="Screenshot 2024-11-28 at 3 31 50 PM" src="https://github.com/user-attachments/assets/e60ba1a7-af58-40f1-a7d9-641e8a7fe12f">

Reply:
but this is a good reminder to check with Wei how exactly this and AC / AF are calculated